src/controller/proxy/controller.go

```diff
@@ -0,0 +1,225 @@
+//  Copyright Project Harbor Authors
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package proxy
+
+import (
+	"context"
+	"fmt"
+	"github.com/docker/distribution"
+	"github.com/docker/distribution/manifest/manifestlist"
+	"github.com/goharbor/harbor/src/common/models"
+	"github.com/goharbor/harbor/src/controller/artifact"
+	"github.com/goharbor/harbor/src/controller/blob"
+	"github.com/goharbor/harbor/src/lib"
+	"github.com/goharbor/harbor/src/lib/errors"
+	"github.com/goharbor/harbor/src/lib/log"
+	"github.com/goharbor/harbor/src/replication/registry"
+	"github.com/opencontainers/go-digest"
+	v1 "github.com/opencontainers/image-spec/specs-go/v1"
+	"io"
+	"net/http"
+	"sync"
+	"time"
+)
+
+const (
+	maxWait             = 10
+	manifestListWaitSec = 900
+	sleepIntervalSec    = 20
+)
+
+var (
+	// Ctl is a global proxy controller instance
+	ctl     Controller
+	ctlLock sync.Mutex
+)
+
+// Controller defines the operations related with pull through proxy
+type Controller interface {
+	// UseLocal check if the manifest should use localHelper
+	UseLocal(ctx context.Context, art lib.ArtifactInfo) bool
+	// ProxyBlob proxy the blob request to the target server
+	ProxyBlob(ctx context.Context, p *models.Project, art lib.ArtifactInfo, w http.ResponseWriter) error
+	// ProxyManifest proxy the manifest to the target server
+	ProxyManifest(ctx context.Context, p *models.Project, art lib.ArtifactInfo, w http.ResponseWriter) error
+}
+type controller struct {
+	blobCtl     blob.Controller
+	registryMgr registry.Manager
+	artifactCtl artifact.Controller
+	local       localInterface
+}
+
+// ControllerInstance -- Get the proxy controller instance
+func ControllerInstance() (Controller, error) {
+	// Lazy load the controller
+	// Because LocalHelper is not ready unless core startup completely
+	ctlLock.Lock()
+	defer ctlLock.Unlock()
+	if ctl == nil {
+		helper, err := newLocalHelper()
+		if err != nil {
+			return nil, err
+		}
+		ctl = &controller{
+			blobCtl:     blob.Ctl,
+			registryMgr: registry.NewDefaultManager(),
+			artifactCtl: artifact.Ctl,
+			local:       helper,
+		}
+	}
+	return ctl, nil
+}
+
+func (c *controller) UseLocal(ctx context.Context, art lib.ArtifactInfo) bool {
+	if len(art.Digest) > 0 {
+		exist, err := c.local.BlobExist(ctx, art.Digest)
```

Comment:
The existence checking logic for manifest and blob are differenet

Reply:
Updated